File: Python/scripts/bst_level_order.py
"""
Create a function level_order to perform a level-order traversal (breadth-first 
search) on a Binary Search Tree. It should print each node's data as 
space-separated integers, traversing each level of the tree from left to right, 
top to bottom. The function takes one parameter, root, which is a pointer to 
the root of the tree. No return value is expected.
"""


class Node:
    def __init__(self, data):
        self.right = self.left = None
        self.data = data


class Solution:
    def insert(self, root, data):
        if root == None:
            return Node(data)
        else:
            if data <= root.data:
                cur = self.insert(root.left, data)
                root.left = cur
            else:
                cur = self.insert(root.right, data)
                root.right = cur
        return root

    def level_order(self, root):
        queue = []
        if root is not None:
            queue.append(root)

        # Continue until the queue is empty
        while len(queue) > 0:
            # Dequeue the front node and print its data
            node = queue.pop(0)
            print(node.data, end=" ")

            # Enqueue the left child
            if node.left is not None:
                queue.append(node.left)

            # Enqueue the right child
            if node.right is not None:
                queue.append(node.right)


T = int(input())
myTree = Solution()
root = None

for i in range(T):
    data = int(input())
    root = myTree.insert(root, data)

myTree.level_order(root)

File: TypeScript/src/bstLevelOrder.ts
class MyNode {
    public left: MyNode | null;
    public right: MyNode | null;
    public data: number;

    constructor(data: number) {
        this.left = null;
        this.right = null;
        this.data = data;
    }
}

class Solution {
    insert(root: MyNode | null, data: number): MyNode {
        if (root === null) {
            return new MyNode(data);
        } else {
            if (data <= root.data) {
                root.left = this.insert(root.left, data);
            } else {
                root.right = this.insert(root.right, data);
            }
        }
        return root;
    }

    levelOrder(root: MyNode | null): void {
        const queue: MyNode[] = [];
        if (root !== null) {
            queue.push(root);
        }

        let output: string = "";
        while (queue.length > 0) {
            const node: MyNode = queue.shift()!;
            output += `${node.data} `;

            if (node.left !== null) {
                queue.push(node.left);
            }
            if (node.right !== null) {
                queue.push(node.right);
            }
        }
        console.log(output.trim());
    }
}

// Example usage with hardcoded inputs
let myTree: Solution = new Solution();
let root: MyNode | null = null;
let data: number[] = [3, 5, 4, 7, 2, 1]; // Example data

data.forEach(d => {
    root = myTree.insert(root, d);
});

myTree.levelOrder(root);
